simpleclient/src/main/java/io/prometheus/client/stream/CKMSStream.java

```diff
@@ -0,0 +1,331 @@
+    /**
+     *  Defines the default targeted-quantiles to track/compute.
+     */
+    private static final Quantile[] DEF_TARGETED_TARGETED_QUANTILEs = new Quantile[]{
+            new Quantile(0.5, 0.05), new Quantile(0.99, 0.001)};
+
+
+    /**
+     *  Invariant function to use based on the type of Quantiles that are being summarized.
+     */
+    private Invariant invariant = null;
+
+    /**
+     *  Tracks the total number of data points that have been observed in the stream.
+     */
+    AtomicInteger count = new AtomicInteger(0);
+
+    /**
+     *  Holds the list of data-points represented by `Sample` at time S(n).
+     */
+    private final LinkedList<Sample> samples = new LinkedList<Sample>();
+
+    /**
+     *  Buffers data-points observed from the data-stream that are yet to be flushed.
+     *
+     *  Reference:
+     *  Incoming items are buffered in sorted order and are inserted/merged
+     *  into the underlying S(n) data-structure
+     */
+    private final ArrayList<T> buffer;
+
+
+    /**
+     *
+     *  @param quantiles The targeted-quantiles that the stream will eventually compute.
+     */
+    public CKMSStream(final Quantile... quantiles) {
+        this.invariant = new TargetedQuantileInvariant(quantiles);
+        this.bufferCap = DEF_BUFFER_CAP;
+        buffer = new ArrayList<T>(bufferCap);
+    }
+
+
+    public CKMSStream() {
+        this(DEF_TARGETED_TARGETED_QUANTILEs);
+    }
+
+    /**
+     *
+     * @param bufferCap The capacity of the buffer. Once reached, a flush is triggered.
+     */
+    public CKMSStream(final int bufferCap) {
+        this(bufferCap, DEF_TARGETED_TARGETED_QUANTILEs);
+    }
+
+    /**
+     *
+     * @param bufferCap The capacity of the buffer. Once reached, a flush is triggered.
+     * @param quantiles The targeted-quantiles that the stream will eventually compute.
+     */
+    public CKMSStream(final int bufferCap, final Quantile... quantiles) {
+        this.invariant = new TargetedQuantileInvariant(quantiles);
+        this.bufferCap = bufferCap;
+        buffer = new ArrayList<T>(bufferCap);
+    }
+
+
+    /**
+     *
+     * @param r The rank of the item in the sample-set
+     * @param n The running count of data-points observed.
+     * @return Delta based on the acceptable error.
+     *
+     */
+    private double f(double r, int n) {
+        double error = invariant.f(r, n);
+        return error;
+    }
+
+
+    /**
+     *
+     * @param v Data point to be observed i.e inserted into our sample-set.
+     */
+    public void insert(final T v) {
+        buffer.add(buffer.size(), v);
+        if (buffer.size() >= bufferCap) {
+            flush();
+        }
+    }
+
+
+    /**
+     *  `merge` merges all data-points from the buffer into the sample-set.
+     *
+     *   It Finds the index that satisfies vi < v ≤ vi+1 and inserts the data-point
+     *   represented by the data-structure e (v, g =1, ∆ = floor(f(ri, n)) − 1)
+     */
+    public void merge() {
```

Comment:
Another note on this method - it's very long. Makes it harder to read.
Consider splitting it